ingenialink/ethercat/network.py

```diff
@@ -455,56 +446,96 @@ def load_firmware(
-            raise ILFirmwareLoadError(
-                f"The firmware file could not be loaded correctly. {foe_return_error}"
-            ) from e
+                error_message += f" Error code: {foe_result}."
+            raise ILFirmwareLoadError(error_message)
+        self.__init_nodes()
         logger.info("Firmware updated successfully")
 
+    def _check_slave_id(self, slave_id: int) -> None:
+        """Check if the slave ID is valid and that it is detected.
+
+        Args:
+            slave_id: The slave ID.
+
+        """
+        if not isinstance(slave_id, int) or slave_id < 0:
+            raise ValueError("Invalid slave ID value")
+        if not self.__is_master_running:
+            self._start_master()
+        if slave_id not in self.__last_init_nodes:
+            self.__init_nodes()
+        if len(self.__last_init_nodes) == 0:
+            raise ILError("Could not find any slaves in the network.")
+        if slave_id not in self.__last_init_nodes:
+            raise ILError(f"Slave {slave_id} was not found.")
+
+    def _switch_to_boot_state(self, slave: "CdefSlave") -> None:
+        """Request the transition to the boot state."""
+        slave.state = pysoem.BOOT_STATE
+        slave.write_state()
+        if (
+            slave.state_check(pysoem.BOOT_STATE, self.ECAT_STATE_CHANGE_TIMEOUT_NS)
+            != pysoem.BOOT_STATE
+        ):
+            raise ILFirmwareLoadError("The drive cannot reach the boot state.")
+
+    def _force_boot_mode(self, slave: "CdefSlave") -> None:
+        """COMOCO drives need to be forced to boot mode."""
+        slave.state = pysoem.PREOP_STATE
+        slave.write_state()
+        if (
+            slave.state_check(pysoem.PREOP_STATE, self.ECAT_STATE_CHANGE_TIMEOUT_NS)
+            == pysoem.PREOP_STATE
+        ):
+            try:
+                slave.sdo_write(
+                    self.__FORCE_COCO_BOOT_IDX,
+                    self.__FORCE_COCO_BOOT_SUBIDX,
+                    self.__FORCE_BOOT_PASSWORD.to_bytes(4, "little"),
+                    False,
+                )
+            except pysoem.WkcError as e:
+                raise ILFirmwareLoadError("Error writing to the Boot mode register.") from e
+        slave.state = pysoem.INIT_STATE
+        slave.write_state()
+        slave.state = pysoem.BOOT_STATE
+        slave.write_state()
+        time.sleep(self.__FORCE_BOOT_SLEEP_TIME_S)
+        self.__init_nodes()
+
+    @staticmethod
+    def _write_foe(slave: "CdefSlave", file_path: str, password: int) -> int:
+        """Write the firmware file via FoE.
+
+        Args:
+            slave: The pysoem slave object.
+            file_path: The firmware file path.
+            password: The firmware password.
+
+        Returns:
+            The FOE operation result.
+
+        """
+        with open(file_path, "rb") as file:
+            file_data = file.read()
+            file_name = os.path.basename(file_path)
```

Comment:
Is it necessary to put the actual file name for the transfer?
I know our bootloaders have a limit in filename size. Not sure if it's documented somewhere

Reply:
The actual file name is irrelevant. We could put any name.